wallet/src/main/scala/org/bitcoins/wallet/sync/WalletSync.scala

```diff
@@ -0,0 +1,85 @@
+package org.bitcoins.wallet.sync
+
+import org.bitcoins.core.protocol.blockchain.{Block, BlockHeader}
+import org.bitcoins.core.util.{BitcoinSLogger, FutureUtil}
+import org.bitcoins.crypto.DoubleSha256DigestBE
+import org.bitcoins.wallet.Wallet
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait WalletSync extends BitcoinSLogger {
+
+  def sync(
+      wallet: Wallet,
+      getBlockHeaderFunc: DoubleSha256DigestBE => Future[BlockHeader],
+      getBestBlockHashFunc: () => Future[DoubleSha256DigestBE],
+      getBlock: DoubleSha256DigestBE => Future[Block])(implicit
+      ec: ExecutionContext): Future[Wallet] = {
+    val bestBlockHashF = getBestBlockHashFunc()
+    val bestBlockHeaderF = for {
+      bestBlockHash <- bestBlockHashF
+      bestheader <- getBlockHeaderFunc(bestBlockHash)
+    } yield bestheader
+
+    val blocksToSyncF = for {
+      bestHeader <- bestBlockHeaderF
+      blocksToSync <- getBlocksToSync(wallet = wallet,
+                                      currentTipBlockHashBE = bestHeader.hashBE,
+                                      accum = Vector.empty,
+                                      getBlock = getBlock)
+    } yield blocksToSync
+
+    val syncedWalletF = for {
+      blocksToSync <- blocksToSyncF
+      syncedWallet <- FutureUtil.foldLeftAsync(wallet, blocksToSync) {
+        case (wallet, nextBlock) =>
+          wallet.processBlock(nextBlock)
+      }
+    } yield syncedWallet
+
+    syncedWalletF
+  }
+
+  /** Syncs the wallet by walking backwards from the currentTip until we reach our wallet's best blockHash */
+  private def getBlocksToSync(
```

Comment:
This function will go forever if the wallet's best block is reorged out. Should probably do some checks with the height.

Reply:
So my thoughts is that this should probably be addressed along with #2179